plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1207,301 +1313,495 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
             (runCostingFunTwoArguments . paramMkCons)
-        where
-          consPlc
-              :: SomeConstant uni a -> SomeConstant uni [a] -> EvaluationResult (Opaque val [a])
-          consPlc
-            (SomeConstant (Some (ValueOf uniA x)))
-            (SomeConstant (Some (ValueOf uniListA xs))) = do
-                DefaultUniList uniA' <- pure uniListA
-                -- Checking that the type of the constant is the same as the type of the elements
-                -- of the unlifted list. Note that there's no way we could enforce this statically
-                -- since in UPLC one can create an ill-typed program that attempts to prepend
-                -- a value of the wrong type to a list.
-                -- Should that rather give us an 'UnliftingError'? For that we need
-                -- https://github.com/input-output-hk/plutus/pull/3035
-                Just Refl <- pure $ uniA `geq` uniA'
-                pure . fromValueOf uniListA $ x : xs
-          {-# INLINE consPlc #-}
+
     toBuiltinMeaning _semvar HeadList =
-        makeBuiltinMeaning
-            headPlc
+        let headListDenotation :: SomeConstant uni [a] -> EvaluationResult (Opaque val a)
+            headListDenotation (SomeConstant (Some (ValueOf uniListA xs))) = do
+                DefaultUniList uniA <- pure uniListA
+                x : _ <- pure xs
+                pure $ fromValueOf uniA x
+            {-# INLINE headListDenotation #-}
+        in makeBuiltinMeaning
+            headListDenotation
             (runCostingFunOneArgument . paramHeadList)
-        where
-          headPlc :: SomeConstant uni [a] -> EvaluationResult (Opaque val a)
-          headPlc (SomeConstant (Some (ValueOf uniListA xs))) = do
-              DefaultUniList uniA <- pure uniListA
-              x : _ <- pure xs
-              pure $ fromValueOf uniA x
-          {-# INLINE headPlc #-}
+
     toBuiltinMeaning _semvar TailList =
-        makeBuiltinMeaning
-            tailPlc
+        let tailListDenotation :: SomeConstant uni [a] -> EvaluationResult (Opaque val [a])
+            tailListDenotation (SomeConstant (Some (ValueOf uniListA xs))) = do
+                DefaultUniList _ <- pure uniListA
+                _ : xs' <- pure xs
+                pure $ fromValueOf uniListA xs'
+            {-# INLINE tailListDenotation #-}
+        in makeBuiltinMeaning
+            tailListDenotation
             (runCostingFunOneArgument . paramTailList)
-        where
-          tailPlc :: SomeConstant uni [a] -> EvaluationResult (Opaque val [a])
-          tailPlc (SomeConstant (Some (ValueOf uniListA xs))) = do
-              DefaultUniList _ <- pure uniListA
-              _ : xs' <- pure xs
-              pure $ fromValueOf uniListA xs'
-          {-# INLINE tailPlc #-}
+
     toBuiltinMeaning _semvar NullList =
-        makeBuiltinMeaning
-            nullPlc
+        let nullListDenotation :: SomeConstant uni [a] -> EvaluationResult Bool
```

Comment:
Maybe we should have called this `IsNullList`.

Reply:
> Maybe we should have called this `IsNullList`.

Too late for that, I guess.

Reply:
> Too late for that, I guess.

(duh, that's what "should have" means Roman, learn some English already)